[Zhang Zhikai] iP

zzhikai opened the pull request:
[](https://nus-cs2103-ay2122s2.github.io/website/schedule/week4/project.html#dukepro)DukePro
“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))


### **Tasks**
- [x] ~~A-Gradle~~

- [x] ~~A-CheckStyle~~

- [x] ~~Level-10~~

- [ ] Fix Styling

Commit:
Add Gradle support

zzhikai committed:
Level 1

zzhikai committed:
Messy code implementation of level 2 Add,List

zzhikai committed:
Minor Fix to Level 2 Add,List

zzhikai committed:
Level 3. Mark as Done

Add mark and unmark features

zzhikai committed:
Level 4. ToDos, Events, Deadline

zzhikai committed:
A-TextUiTesting

zzhikai committed:
A-Exceptions

zzhikai committed:
Level-6 Delete

zzhikai committed:
Add enums

zzhikai committed:
Change to enum method

zzhikai committed:
Enum method and edit testing file

zzhikai committed:
Clean up code

zzhikai committed:
Level 7. Save

zzhikai committed:
Add date into deadline

Able to parse 4 date formats

zzhikai committed:
Edit exception message for date formatting

zzhikai committed:
Add empty description check for deadline and event

To implement event date handling

zzhikai committed:
Merge pull request #1 from zzhikai/branch-Level-7

Level 7. Save

zzhikai committed:
Merge branch 'master' into branch-Level-8

zzhikai committed:
Merge pull request #2 from zzhikai/branch-Level-8

Level 8. Dates and Times

zzhikai committed:
Remove merge conflict markers

zzhikai committed:
Create TaskList class for OOP

zzhikai committed:
Add Storage and Ui class

zzhikai committed:
Refactor code

Create different packages and did more OOP

zzhikai committed:
Clean up code

zzhikai committed:
Clean up code

zzhikai committed:
Try JUnit test without gradle

zzhikai committed:
Separate Tests

zzhikai committed:
Remove debugging print statements

zzhikai committed:
Add javadocs for classes in Task package

zzhikai committed:
Add javadocs for TaskList class

zzhikai committed:
Add JavaDocs for classes in Exception package

zzhikai committed:
Add JavaDocs for classes in Command package

zzhikai committed:
Fix typo and clean up JavaDocs

zzhikai committed:
Add JavaDocs for Parser class

zzhikai committed:
Edit to adhere to code standards

zzhikai committed:
Level 9. Find

zzhikai committed:
Merge pull request #3 from zzhikai/A-JavaDoc

Add JavaDocs

zzhikai committed:
Merge pull request #4 from zzhikai/A-CodingStandard

Edit Code to Adhere to Coding Standard

zzhikai committed:
Merge branch 'master' into Level-9

zzhikai committed:
Merge pull request #5 from zzhikai/Level-9

Level 9. Find

zzhikai committed:
Add JavaDocs for new method

zzhikai committed:
Adhere to reuse policy

zzhikai committed:
Change to own github username

DavidTan0527 commented on src/main/java/Duke/Task/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+                    formatToParse = DateTimeFormatter.ofPattern(dateFormats.get(i));
+                    this.date = LocalDate.parse(by, formatToParse);
+                    break;
+                } catch (Exception wrongFormat) {
```
This will also capture the Exception at line 31 when `i == dateFormats.size()`.
Perhaps, it might be a better practice to put the `if` statement at line 36 before `try...catch`, and catch only `DateTimeParseException`.

zzhikai replied:
Thanks for the suggestion. 
What I was intending to do here is that because the LocalDate.parse will always return exception if the format used is incorrect, I needed to catch the Exception to continue to try the next format, only when all formats are exhausted, I throw the exception. 
Hence its kind of difficult to put the statement outside(I have to admit this solution is kind of hard coded as well)😅

DavidTan0527 commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,84 @@
+package Duke;
+
+import Duke.Command.*;
+import Duke.Parser.Parser;
+import Duke.Storage.Storage;
+import Duke.Task.*;
+import Duke.Exception.*;
+import Duke.TaskList.TaskList;
+import Duke.Ui.Ui;
+
+import java.io.*;
```
Avoid wildcard imports. It might be better to explicitly import the packages you're using.

DavidTan0527 commented on src/main/java/Duke/Parser/Parser.java:
```diff
@@ -0,0 +1,59 @@
+package Duke.Parser;
+
+import Duke.Command.*;
```
Avoid wildcard imports. It might be better to explicitly import the packages you're using.

DavidTan0527 commented on src/main/java/Duke/Storage/Storage.java:
```diff
@@ -0,0 +1,47 @@
+
+import Duke.Task.Task;
+
+import java.io.*;
```
Avoid wildcard imports. It might be better to explicitly import the packages you're using.

DavidTan0527 commented on src/main/java/Duke/Task/Event.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Represents an event task.
+ */
+public class Event extends Task{
```
```suggestion
public class Event extends Task {
```

DavidTan0527 commented on src/main/java/Duke/TaskList/TaskList.java:
```diff
@@ -0,0 +1,150 @@
+package Duke.TaskList;
+
+import Duke.Task.Task;
+import Duke.Exception.*;
```
Avoid wildcard imports.
```suggestion
import Duke.Exception.DukeException;
```

DavidTan0527 commented on src/main/java/Duke/TaskList/TaskList.java:
```diff
@@ -0,0 +1,150 @@
+        if (searchResult.size() == 0) {
+            throw new EmptyListException("No results found!");
+        }
+        for (int i = 0; i < searchResult.size(); i++) {
+            System.out.println((i + 1) + "." + searchResult.get(i).toString());
+        }
```
This is quite similar to `printOutTaskList`. Maybe we can create a `TaskList` here after filtering, then call the method as to not repeat the code.

zzhikai replied:
Thank you for this suggestion! Makes the code a lot more efficient and clean👍  8475c55b322b6e5013a713391e12e80d7f5378e9

DavidTan0527 reviewed (commented):
Overall great code! Just a few coding standard issues and some suggestions that you can hopefully look into. 😄

zzhikai committed:
Merge branch 'add-gradle-support' of https://github.com/zzhikai/ip into add-gradle-support

zzhikai committed:
A-CheckStyle

zzhikai committed:
Merge pull request #6 from zzhikai/add-gradle-support

Add gradle support

zzhikai committed:
Merge branch 'master' into A-CheckStyle

zzhikai committed:
Merge pull request #7 from zzhikai/A-CheckStyle

A check style

zzhikai committed:
Rename packages to lowercase

zzhikai committed:
Use tasklist to print out find results instead of manually coding

yusufaine commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,88 @@
+
+public class Duke {
+
+    private static Ui ui;
+    private Ui uiDuke;
```
Was `uiDuke` used anywhere else?

yusufaine commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,88 @@
+        taskDataList = new TaskList(inputDatabase);
+    }
+
+    private static void printFileContents(String filePath) throws FileNotFoundException {
+        File f = new File(filePath);
+        Scanner s = new Scanner(f);
+        while (s.hasNext()) {
+            System.out.println(s.nextLine());
+        }
+
+    }
```
I'm curious, what was your rationale for placing this command here? 
I personally would consider moving this into `Command` or `Ui`?

yusufaine commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,88 @@
+        chatBot.run();
+    }
+
+    public static void writeDataInputToDisk(ArrayList<Task> inputTaskList) {
+        try {
+            FileOutputStream writeDatabaseInput = new FileOutputStream("src/main/java/TaskDatabase.ser");
+            ObjectOutputStream writeDatabaseStream = new ObjectOutputStream(writeDatabaseInput);
+            writeDatabaseStream.writeObject(inputTaskList);
+            writeDatabaseStream.flush();
+            writeDatabaseStream.close();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
```
Similar to `printFileContents()`

yusufaine commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,88 @@
+        }
+    }
+
+    public static int indexOfTask(String command) {
+        int index = command.length() + 1;
+        return index;
+    }
+
+    public static int endIndexOfTask(String input) {
+        int index = input.indexOf("/");
+        return index;
+    }
```
Consider adding JavaDocs to explain this portion as I don't quite understand what this methods are used for, sorry 😅

yusufaine commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,43 @@
+        if (description == "") {
+            throw new EmptyDescriptionException("OOPS!!! The description cannot be empty.");
+        }
+        try {
+            int taskNumber = Integer.valueOf(description);
+            taskIndex = taskNumber - 1;
+        } catch (NumberFormatException e) {
+            throw new DukeException("Please input number after unmark");
+        }
```
Errors that you may consider checking for:
1. Invalid number, and
2. Index out of bounds

yusufaine commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,51 @@
+     *
+     * @param taskList TaskList for command to add task to.
+     */
+    @Override
+    public void callCommand(TaskList taskList) {
+        taskList.addTask(this.task);
+    }
```
Not really commenting about your code here, but I like how you're handling the execution of the command, I might modify mine to make it like this as it makes more sense! 👍

yusufaine commented on src/main/java/duke/exception/EmptyByException.java:
```diff
@@ -0,0 +1,10 @@
+package duke.exception;
```
I think this is a good way to handle exceptions as the class name is self-explanatory. 👍

yusufaine commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,59 @@
+        String inputBody = inputStrings.length == 2 ? inputStrings[1] : "";
+        switch (inputCommand) {
+        case "list":
+            return new ListCommand("");
```
more nit-picky then anything, I find it a little odd a new object is initialised with an empty string  but don't worry, I understand that the `Command` parent class required you to give a parameter 😆

yusufaine commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+
+    //@@author zzhikai-reused
+    //Reused from https://samderlust.com/dev-blog/java/write-read-arraylist-object-file-java
+    // with minor modification
```
Was there any particular reason why you used his code instead of something simpler like `FileReader`/`FileWriter`? 🤔

yusufaine commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,18 @@
+package duke.ui;
+
+public class Ui {
```
Maybe you can consider bringing the user input/output here since `UI`  is supposed to "deals with interactions with the user". I personally made all my print methods and `Scanner` in this class so that the I/O stuff is isolated here.

yusufaine reviewed (commented):
Overall rather clean code layout, just a few nit-picky stuff here and there so don't take them too seriously, it's mainly because we've done our code a little differently so I was curious as to why you did things a certain way and left some suggestions (that you can choose to (not) implement 😆 )

zzhikai committed:
Level 10. GUI

zzhikai committed:
Merge pull request #8 from zzhikai/Level-10

Level 10. GUI

zzhikai committed:
Modified build gradle for jar file

zzhikai committed:
Add assertions for Mark and Unmark in TaskList

zzhikai committed:
Add handling after finding error from Assertion

zzhikai committed:
Functions not being used, uneccesary comments found

Removed some unused functions and comments

Let's,
* Remove all unused code and redundant comments
* Fix Styling errors

zzhikai committed:
Merge pull request #9 from zzhikai/A-Assertions

A-Assertions A-FullCommitMessage

zzhikai committed:
Code quality was poor, did not pass checkstyle

Formatted coding style of intellij in accordance to guide given.

Configured coding style with:
https://se-education.org/guides/tutorials/intellijCodeStyle.html

zzhikai committed:
A-CodeQuality changes was pushed in master branch

Created this A-CodeQuality branch to keep track of that information

Note that changes supposedly done in this branch is already in the previous commit before this PR.

zzhikai committed:
Merge pull request #10 from zzhikai/A-CodeQuality

A-CodeQuality changes was pushed in master branch

zzhikai committed:
Find task method not using streams

Attempted using Stream Method to collect into a new ArrayList

findTask Method now uses Streams

zzhikai committed:
Merge pull request #11 from zzhikai/A-Streams

Find task method not using streams

zzhikai committed:
Add github actions

zzhikai committed:
Merge pull request #12 from zzhikai/A-CI

A-CI

zzhikai committed:
Remove old code

zzhikai committed:
Create main.yml

zzhikai committed:
Merge pull request #13 from zzhikai/A-testGithubActions

Remove old code

zzhikai committed:
Remove IO rediretion test from main github workflow

zzhikai committed:
Merge pull request #14 from zzhikai/zzhikai-github-actions

Create main.yml

Fix Github Actions by removing IO redirection test.

zzhikai committed:
Fix bug in delete command

Method was trying to get task toString after deleting it

zzhikai committed:
To implement C-Update

Implemented update feature to edit the description of different task

Command is: update {Task Number} {New Description}

Add method within Task to update description of each task

TaskList gets task and calls on task method to update description.

zzhikai committed:
Merge pull request #15 from zzhikai/C-Update

To implement C-Update

zzhikai committed:
Improve GUI

zzhikai committed:
A-UserGuide

zzhikai committed:
Merge pull request #18 from zzhikai/A-BetterGui

Improve GUI

zzhikai committed:
Merge pull request #17 from zzhikai/A-UserGuide

A-UserGuide

zzhikai committed:
Set theme jekyll-theme-slate

zzhikai committed:
Update User Guide

zzhikai committed:
Set theme jekyll-theme-minimal

zzhikai committed:
Set theme jekyll-theme-cayman

zzhikai committed:
Set theme jekyll-theme-cayman

zzhikai committed:
Set theme jekyll-theme-hacker

zzhikai committed:
Remove unused code

zzhikai committed:
Merge branch 'master' of https://github.com/zzhikai/ip

zzhikai committed:
Update date handling in event command

zzhikai committed:
Move storing of database file

zzhikai committed:
Update release

zzhikai committed:
Update Readme

zzhikai committed:
Minor fix to code quality

zzhikai committed:
Fix javadocs, Minor changes to MainWindow using scenebuilder

zzhikai committed:
Removed unused code

zzhikai committed:
Merge pull request #19 from zzhikai/fix-code-quality

Minor fix to code quality

zzhikai committed:
fix code quality, remove dead code

zzhikai committed:
Merge pull request #21 from zzhikai/fix-quality

fix code quality, remove dead code
